### ISSUE TITLE
[bug] TypescriptApp: Release task runs git diff in the wrong context

### ISSUE THREAD

**Issue by alexforsyth** [projen/projen#1217]
This should be a fairly minor bug.

## Issue
The release task runs this command:
```
👾 release | git diff --ignore-space-at-eol --exit-code
```
However, if the project project is not at the root of the git repo, it fails for unrelated diffs.

## Fix
 The release task should run this command:
```
👾 release | git diff ./ --ignore-space-at-eol --exit-code
```
This should solve both the case where it's running at the root of the project and not at the root of the project.

**Comment by moltar** [comment]
https://drewdevault.com/2021/10/26/stalebot.html

**Comment by cfuerst** [comment]
An example can be observed here after chaining the readme in main: 

https://github.com/cloudy-with-a-chance-of-meatballs/cdk-lambda-token-authorizer-jwt/actions/runs/3448987353/jobs/5756509741

**Comment by pchaganti** [comment]
Still seeing this issue. Any idea how to resolve this?

thanks!

**Comment by parraletz** [comment]
I'm having the issue, Do you have Any idea how to resolve it?

**Comment by parraletz** [comment]
Hey guys I've worked in a workaround, in my case I'm working on a cdk construct, and I've updated release step in projen.ts

project.tasks
  .tryFind('release')
  ?.updateStep(4, { exec: 'git diff --ignore-space-at-eol --exit-code | tee' })


https://github.com/occmundial/occ-ecr-pattern/blob/ed6bdf8c66caf1437819e2dfe90d6f43620aaaa4/.projenrc.ts#L40

The issue is because the exit code from git diff is 1 and tee change to 0.

**Comment by mrgrain** [comment]
> Hey guys I've worked in a workaround, in my case I'm working on a cdk construct, and I've updated release step in projen.ts
> 
> project.tasks .tryFind('release') ?.updateStep(4, { exec: 'git diff --ignore-space-at-eol --exit-code | tee' })

For anyone else reading, **this is NOT a good idea**. This check exists for a reason. It is part of anti tampering measures and ensures that code has not been changed during the release process.

> The issue is because the exit code from git diff is 1 and tee change to 0.

Why do you need to wholesale ignore any changes in your repo after running release? The OP issue is about the release task not differentiating between root and subprojects.

**Comment by mrgrain** [comment]
@alexforsyth I know this has been ages. This came up in a similar context for the build of subprojects. Build also use this command for anti-tamper protection and self-mutation. When discussing it, we arrived at preferring the mutation check to always include the _full repo_. The reason being that a subproject should never change its parent project anyway, and if it does it should definitely treated the same way as any other anti-tamper checks. In your original issue, you stated this is a bug. But can you elaborate why?

Long story short, I currently believe we should **not** change the behavior. And checking for changes in the whole repo is indeed a feature, not a bug.

**Comment by hakenmt** [comment]
I'm not sure if my problem is related or if I'm just doing it wrong (and there's a high likelihood that could be the case), but git diff complains about every commit I make. Change a space in a file, commit, push, and the the git diff identifies a difference between the old file without the space and the new file with the space. It also complains about every auto-formatting correction that gets made to the `.projenrc.ts` file, for example, if I don't have a comma after the last entry in an array (which is normal), the auto-formatting that happens somewhere on the `.projenrc.ts` file adds that comma in places, and then the git diff fails. All I can say is that I find it frustrating to the point that I'd want it off, every commit takes two tries to make sure I match whatever the build in Github actions is producing. This also means that I have to constantly build locally first before checking in to get files to match.

**Comment by mrgrain** [comment]
Thanks for your feedback @hakenmt  I can understand your frustrations. Unfortunately it would be quite difficult to distinguish between 'unimportant' formatting changes and potentially relevant tampering attempts. We should probably make this an explicitly configurable option though.

> but git diff complains about every commit I make

This should only happen on PRs and releases, not on every commit. (Though I guess on a PR it happens on every commit.)

> Change a space in a file, commit, push, and the the git diff identifies a difference between the old file without the space and the new file with the space.
> It also complains about every auto-formatting correction that gets made to the .projenrc.ts file, for example, if I don't have a comma after the last entry in an array (which is normal), the auto-formatting that happens somewhere on the .projenrc.ts file adds that comma in places, and then the git diff fails.

This very strongly hints that the formatting & linting configurations you have defined in projen (or have been defined for you) are not aligned with your local settings, e.g. in your IDE. Thinks like commas after the last entry in an array are entirely subjective. You should define your preferences in the rules for your project and enforce them consistently. It sounds like you are just not agreeing with the default rules, and should change them. Or if you don't really mind either way, you might have to configure your IDE to follow the rules defined for the project.

> This also means that I have to constantly build locally first before checking in to get files to match.

Yes, it is certainly the expectation that projects are build and tested locally before being pushed. Depending on the project type, you might get away with running `npx projen eslint` or `npx projen lint`.

**Comment by spensireli** [comment]
I had this exact issue and i forgot to run `npx projen build` before doing my commit to git. Doing so resolved the issue.

**Comment by mrgrain** [comment]
Closing as our current stance is that the tampering check should always include the whole repository. If you disagree, please open a new issue.

If you are here because your workflow failed at `git diff --ignore-space-at-eol --exit-code` and a search led you here, please see the first post.
